File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
    void solve(vector<int>& candidates, int target, vector<vector<int>> &answer, vector<int> temp, int i)
    {
        if(target==0)
        {
            answer.push_back(temp);
            return;
        }
        if(i==candidates.size())
        {
            return;
        }
        if(target>=candidates[i])
        {
            temp.push_back(candidates[i]);
            solve(candidates,target-candidates[i],answer,temp,i+1);
            temp.pop_back();
            while(i+1<candidates.size() && candidates[i]==candidates[i+1])
            {
                i++;
            }
        }
        solve(candidates,target,answer,temp,i+1);
    }
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(),candidates.end());
        vector<vector<int>> answer;
        vector<int> temp;
        solve(candidates,target,answer,temp,0);
        return answer;
    }
};

File: 0040-combination-sum-ii/0040-combination-sum-ii.java
class Solution {
    private void solve(int[] candidates, int target, List<List<Integer>>answer, List<Integer>temp, int i)
    {
        if(target==0)
        {
            answer.add(new ArrayList<>(temp));
            return;
        }
        if(i==candidates.length)
        {
            return;
        }
        if(target>=candidates[i])
        {
            temp.add(candidates[i]);
            solve(candidates,target-candidates[i],answer,temp,i+1);
            temp.remove(temp.size()-1);
            while(i+1<candidates.length && candidates[i]==candidates[i+1])
            {
                i++;
            }
        }
        solve(candidates,target,answer,temp,i+1);
    }
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> answer=new ArrayList<>();
        List<Integer> temp=new ArrayList<>();
        solve(candidates,target,answer,temp,0);
        return answer;
    }
}